### PR TITLE
Add data-histograms endpoint and checksum_histogram replicastore method

### PR DESCRIPTION
**Pull request by gravityrail**
Create an endpoint which can be used to compare buckets of checksums across replicastores, as a way of drilling down into disrepancies and ultimately performing selective sync

On WPCOM, invoking this API looks like this:

```
wp> $validator->get_site_histogram( 'posts', 5 );
array(15) {
  '1-51' =>
  string(8) "2ADAE1EC"
  '52-90' =>
  string(8) "5B073436"
  '91-129' =>
  string(8) "B8AD19B7"
  '130-168' =>
  string(8) "A927F6CB"
  '169-207' =>
  string(7) "60028C6"
  '208-246' =>
  string(8) "6FDB19B8"
  '247-285' =>
  string(8) "99A33198"
  '286-324' =>
  string(8) "85329053"
  '325-363' =>
  string(8) "2C9A485F"
  '364-402' =>
  string(8) "C69DEE10"
  '403-441' =>
  string(8) "CD9C1A9F"
  '442-480' =>
  string(8) "105F977E"
  '481-519' =>
  string(8) "59B764F4"
  '520-558' =>
  string(8) "97AD1E3A"
  '559-600' =>
  string(8) "A8C759DA"
}
```

This PR also adds some unrelated functionality: sync queue size and lag for the sync status API

cc @lezama

### CODE REVIEW

File: json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php [hunk 1]
```diff
@@ -89,6 +89,45 @@ protected function result() {
 	}
 }
 
+// GET /sites/%s/data-histogram
+class Jetpack_JSON_API_Sync_Histogram_Endpoint extends Jetpack_JSON_API_Endpoint {
+	// GET /sites/%s/cached-data-check
+	protected $needed_capabilities = 'manage_options';
```
Comment (lezama) [inline json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php]: oops :)

File: tests/php/sync/test_interface.jetpack-sync-replicastore.php [hunk 2]
```diff
@@ -110,6 +110,23 @@ function get_all_checksums( $replicastore ) {
...
+	 * @dataProvider store_provider
+	 * @requires PHP 5.3
+	 */
+	function test_checksum_histogram( $store ) {
+
+		$post = self::$factory->post( 5 );
+
+		$histogram = $store->checksum_histogram( 'posts', 10, 0, 0 );
+
+		error_log(print_r($histogram, 1));
```
Comment (lezama) [inline tests/php/sync/test_interface.jetpack-sync-replicastore.php]: :)

File: json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php [hunk 3]
```diff
@@ -89,6 +89,45 @@ protected function result() {
...
+		}
+
+		if ( is_wp_error( $result ) ) {
+			$sync_queue->unlock();
+			return $result;
+		}
+
+		$args = $this->query_args();
+
+		require_once dirname(__FILE__).'/../../sync/class.jetpack-sync-wp-replicastore.php';
```
Comment (lezama) [inline json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php]: we need to add spaces for WP standards around ` . `

File: json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php [hunk 4]
```diff
@@ -89,6 +89,45 @@ protected function result() {
 	}
 }
 
+// GET /sites/%s/data-histogram
+class Jetpack_JSON_API_Sync_Histogram_Endpoint extends Jetpack_JSON_API_Endpoint {
+	// GET /sites/%s/cached-data-check
+	protected $needed_capabilities = 'manage_options';
+
+	protected function result() {
+		require_once dirname(__FILE__).'/../../sync/class.jetpack-sync-sender.php';
```
Comment (lezama) [inline json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php]: we need to add spaces for WP standards around ` . `

**Comment by lezama** [comment]
just saw it is progress, ignore my comments, looks good so far :)

File: sync/class.jetpack-sync-wp-replicastore.php [hunk 5]
```diff
@@ -39,12 +39,30 @@ function full_sync_end( $checksum ) {
 		// noop right now
 	}
 
-	public function post_count( $status = null ) {
-		return count( $this->get_posts( $status ) );
+	public function post_count( $status = null, $min_id = null, $max_id = null ) {
+		global $wpdb;
+
+		$where = "";
```
Comment (lezama) [inline sync/class.jetpack-sync-wp-replicastore.php]: we could use single quotes for strings inside this method

**Comment by lezama** [comment]
LGTM, I haven't tested it exhaustively but code looks good(except for some minor lint errors that we can fix in another pass).

File: sync/class.jetpack-sync-wp-replicastore.php [hunk 6]
```diff
@@ -566,6 +615,65 @@ public function checksum_all() {
...
+			$where .= " AND $id_field >= " . intval( $start_id );
+		}
+
+		if ( $end_id ) {
+			$where .= " AND $id_field <= " . intval( $end_id );
+		}
+
+		do {
+			list( $first_id, $last_id ) = $wpdb->get_row( 
+				"SELECT MIN($id_field) as min_id, MAX($id_field) as max_id FROM ( SELECT $id_field FROM $object_table WHERE $where ORDER BY $id_field ASC LIMIT $query_offset, $bucket_size ) as ids",
```
Comment (lezama) [inline sync/class.jetpack-sync-wp-replicastore.php]: is `as ids` needed?

Comment (gravityrail) [inline sync/class.jetpack-sync-wp-replicastore.php]: yes, MySQL requires an alias for subqueries.

**Pull request merged**